### PR TITLE
Audit Refactoring #1542

### PR DESCRIPTION
**Pull request by prb112**
- Refactor to separate Configuration, Mapping and AuditService
- Configuration to use EventStreams Binding by default
- Add a MapperFactory to split between cadf and auditevent types
	- AuditEventMapper - Wraps the AuditLogEntry into AuditEvent format
	- CADFMapper - Wraps the AuditLogEntry into CADF format
- Audit Service is now KafkaService, NoOpService separating the
Streaming framework from the mapping actions.
- Removed duplicate code for Event/EventStrams processing
- Add Test Coverage for Uncovered Classes

To check 
```bash bin/kafka-console-consumer.sh --bootstrap-server my-broker-1:9093,my-broker-2:9093 --topic FHIR_AUDIT --max-messages 25 --property print.timestamp=true --consumer.config client-ssl.properties```

# AuditEvent
Example AuditEvent output of query `127.0.0.1/localhost/{_tag=[http://ibm.com/fhir/tag|tag2,http://ibm.com/fhir/tag|tag], _count=[1000], _page=[1]}`

formatted... 

```
{
    "resourceType": "AuditEvent",
    "type": {
        "system": "http://terminology.hl7.org/CodeSystem/audit-event-type",
        "code": "rest",
        "display": "Restful Operation"
    },
    "subtype": [
        {
            "system": "http://hl7.org/fhir/restful-interaction",
            "code": "search",
            "display": "search"
        }
    ],
    "action": "E",
    "period": {
        "start": "2020-12-10T16:38:14.466Z",
        "end": "2020-12-10T16:38:14.631Z"
    },
    "recorded": "2020-12-10T11:38:14.632173-05:00",
    "outcome": "0",
    "outcomeDesc": "success",
    "purposeOfEvent": [
        {
            "coding": [
                {
                    "system": "http://terminology.hl7.org/CodeSystem/v3-ActReason",
                    "code": "PurposeOfUse",
                    "display": "PurposeOfUse"
                }
            ]
        }
    ],
    "agent": [
        {
            "role": [
                {
                    "coding": [
                        {
                            "system": "http://terminology.hl7.org/CodeSystem/extra-security-role-type",
                            "code": "datacollector",
                            "display": "datacollector"
                        }
                    ]
                }
            ],
            "name": "fhir-server",
            "requestor": true,
            "network": {
                "address": "internal-ip",
                "type": "1"
            }
        }
    ],
    "source": {
        "site": "127.0.0.1/localhost",
        "observer": {
            "reference": "fhir-server"
        },
        "type": [
            {
                "system": "http://terminology.hl7.org/CodeSystem/security-source-type",
                "code": "4",
                "display": "Application Server"
            }
        ]
    },
    "entity": [
        {
            "securityLabel": [
                {
                    "system": "http://terminology.hl7.org/CodeSystem/v3-Confidentiality",
                    "code": "N",
                    "display": "normal"
                }
            ],
            "description": "FHIR Search request",
            "query": "MTI3LjAuMC4xL2xvY2FsaG9zdC97X3RhZz1baHR0cDovL2libS5jb20vZmhpci90YWd8dGFnMixodHRwOi8vaWJtLmNvbS9maGlyL3RhZ3x0YWddLCBfY291bnQ9WzEwMDBdLCBfcGFnZT1bMV19",
            "detail": [
                {
                    "type": "FHIR Context",
                    "valueBase64Binary": "CnsKICAgICJyZXF1ZXN0X3VuaXF1ZV9pZCI6ICIzMmJmMDNhNS1kNmQ1LTQ2MTEtYmFjYS0wNjdkMzcwMDYyMzUiLAogICAgImFjdGlvbiI6ICJSIiwKICAgICJzdGFydF90aW1lIjogIjIwMjAtMTItMTAgMTY6Mzg6MTQuNDY2IiwKICAgICJlbmRfdGltZSI6ICIyMDIwLTEyLTEwIDE2OjM4OjE0LjYzMSIsCiAgICAiYXBpX3BhcmFtZXRlcnMiOiB7CiAgICAgICAgInJlcXVlc3QiOiAiaHR0cHM6Ly9sb2NhbGhvc3Q6OTQ0My9maGlyLXNlcnZlci9hcGkvdjQvX3NlYXJjaD9fdGFnPWh0dHAlM0ElMkYlMkZpYm0uY29tJTJGZmhpciUyRnRhZyU3Q3RhZzIlMkNodHRwJTNBJTJGJTJGaWJtLmNvbSUyRmZoaXIlMkZ0YWclN0N0YWcmX2NvdW50PTEwMDAmX3BhZ2U9MSIsCiAgICAgICAgInJlcXVlc3Rfc3RhdHVzIjogMjAwCiAgICB9LAogICAgInF1ZXJ5IjogIntfdGFnPVtodHRwOi8vaWJtLmNvbS9maGlyL3RhZ3x0YWcyLGh0dHA6Ly9pYm0uY29tL2ZoaXIvdGFnfHRhZ10sIF9jb3VudD1bMTAwMF0sIF9wYWdlPVsxXX0iLAogICAgImJhdGNoIjogewogICAgICAgICJyZXNvdXJjZXNfcmVhZCI6IDIKICAgIH0sCiAgICAiZXZlbnRfdHlwZSI6ICJmaGlyLXNlYXJjaCIsCiAgICAiZGVzY3JpcHRpb24iOiAiRkhJUiBTZWFyY2ggcmVxdWVzdCIsCiAgICAibG9jYXRpb24iOiAiMTI3LjAuMC4xL2xvY2FsaG9zdCIKfQ=="
                }
            ]
        }
    ]
}
```

# CADF Event

```
{
    "action": "create",
    "eventTime": "2020-12-10 16:49:22.307",
    "eventType": "activity",
    "id": "c62ee8f8-be77-49d0-aad0-c0bf52a05fdf",
    "outcome": "success",
    "typeURI": "http://schemas.dmtf.org/cloud/audit/1.0/event",
    "tags": [
    ],
    "attachments": [
        {
            "contentType": "application/json",
            "content": "rO0ABXQCUQp7CiAgICAicmVxdWVzdF91bmlxdWVfaWQiOiAiYzYyZWU4ZjgtYmU3Ny00OWQwLWFhZDAtYzBiZjUyYTA1ZmRmIiwKICAgICJhY3Rpb24iOiAiQyIsCiAgICAic3RhcnRfdGltZSI6ICIyMDIwLTEyLTEwIDE2OjQ5OjIyLjE2OCIsCiAgICAiZW5kX3RpbWUiOiAiMjAyMC0xMi0xMCAxNjo0OToyMi4zMDciLAogICAgImFwaV9wYXJhbWV0ZXJzIjogewogICAgICAgICJyZXF1ZXN0IjogImh0dHBzOi8vbG9jYWxob3N0Ojk0NDMvZmhpci1zZXJ2ZXIvYXBpL3Y0L1ZlcmlmaWNhdGlvblJlc3VsdCIsCiAgICAgICAgInJlcXVlc3Rfc3RhdHVzIjogMjAxCiAgICB9LAogICAgImRhdGEiOiB7CiAgICAgICAgInJlc291cmNlX3R5cGUiOiAiVmVyaWZpY2F0aW9uUmVzdWx0IiwKICAgICAgICAiaWQiOiAiMTc2NGQ4ZWEzMGEtYzUxOGMxOWItMWM0Zi00Yjc0LThhMDMtNGM5NjA0YmZkYjZlIiwKICAgICAgICAidmVyc2lvbl9pZCI6ICIxIgogICAgfSwKICAgICJldmVudF90eXBlIjogImZoaXItY3JlYXRlIiwKICAgICJkZXNjcmlwdGlvbiI6ICJGSElSIENyZWF0ZSByZXF1ZXN0IiwKICAgICJsb2NhdGlvbiI6ICIxMjcuMC4wLjEvbG9jYWxob3N0Igp9"
        }
    ],
    "initiator": {
        "id": "default@fhir-server",
        "typeURI": "compute/machine",
        "host": "192.168.86.20",
        "credential": {
            "token": "user-fhiruser"
        },
        "geolocation": {
            "city": "Dallas",
            "state": "TX",
            "region": "US",
            "annotations": [
            ]
        }
    },
    "observer": {
        "id": "fhir-server",
        "typeURI": "compute/node",
        "name": "IBM FHIR Server - Audit",
        "geolocation": {
            "city": "Dallas",
            "state": "TX",
            "region": "US",
            "annotations": [
            ]
        }
    }
}
```

Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

### CODE REVIEW

File: docs/src/pages/guides/FHIRServerUsersGuide.md [hunk 1]
```diff
@@ -1778,10 +1810,180 @@ And then in the YAML file for your Kubernetes deployment, specify the environmen
...
+            "ssl.protocol": "TLSv1.2",
+            "ssl.enabled.protocols": "TLSv1.2",
+            "ssl.endpoint.identification.algorithm": "HTTPS"
+        },
+        "kafkaServers": "********",
+        "kafkaApiKey": "********"
+    }
+```
+
+The service may be map to the CADF format or the FHIR AuditEvent resource format by declaring a mapper type - 'cadf' or 'auditevent'.
```
Comment (lmsurpre) [inline docs/src/pages/guides/FHIRServerUsersGuide.md:1842]:
```suggestion
The service can map to the CADF format or the FHIR AuditEvent resource format by declaring a mapper type - 'cadf' or 'auditevent'. 
```

Comment (lmsurpre) [inline docs/src/pages/guides/FHIRServerUsersGuide.md:1842]: I would have assumed the AuditEvent mapper would just be a different class.  If WhcAuditCadfLogService is used to map to both CADF and FHIR, then I think it should get renamed.

Comment (prb112) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:1842]: Discussed as a team on code review

File: docs/src/pages/guides/FHIRServerUsersGuide.md [hunk 2]
```diff
@@ -1778,10 +1810,180 @@ And then in the YAML file for your Kubernetes deployment, specify the environmen
...
+    ],
+    "action": "E",
+    "period": {
+        "start": "2020-12-10T16:38:14.466Z",
+        "end": "2020-12-10T16:38:14.631Z"
+    },
+    "recorded": "2020-12-10T11:38:14.632173-05:00",
+    "outcome": "0",
+    "outcomeDesc": "success",
+    "purposeOfEvent": [
```
Comment (lmsurpre) [inline docs/src/pages/guides/FHIRServerUsersGuide.md:1916]: is there supposed to be an actual purpose string in here somewhere.  having a "purpose" of "PurposeOfUse" seems pretty odd

Comment (lmsurpre) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:1916]: Interesting, apparently that *is* a valid value in the ValueSet.  Its a top-level code which has lots of potential children.  And its the *only* top-level code in the valueset.  So weird.

File: fhir-audit/pom.xml [hunk 3]
```diff
@@ -28,10 +33,6 @@
             <artifactId>testng</artifactId>
             <scope>test</scope>
         </dependency>
-        <dependency>
-            <groupId>com.google.code.gson</groupId>
-            <artifactId>gson</artifactId>
```
Comment (lmsurpre) [inline fhir-audit/pom.xml:33]: yay!  down to just 2 or 3 json serialization libs :-)

File: fhir-audit/src/main/java/com/ibm/fhir/audit/logging/api/AuditLogServiceFactory.java [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*
- * (C) Copyright IBM Corp. 2016,2019
+ * (C) Copyright IBM Corp. 2016,2020
```
Comment (lmsurpre) [inline fhir-audit/src/main/java/com/ibm/fhir/audit/logging/api/AuditLogServiceFactory.java:2]: super-duper minor, but I usually put a space after the `,` as i edit it (a change from what we did when we auto-added these when first going open)

File: fhir-audit/src/main/java/com/ibm/fhir/audit/logging/api/configuration/ConfigurationTranslator.java [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+            props.put("ssl.enabled.protocols", "TLSv1.2");
+        }
+        if(!props.containsKey("ssl.endpoint.identification.algorithm")) {
+            props.put("ssl.endpoint.identification.algorithm", "HTTPS");
+        }
+    }
+
+    /**
+     * gets the location to load the properties from:
+     * 1 - Environment Bindings with EventStreams Credentials format - "env"
```
Comment (lmsurpre) [inline fhir-audit/src/main/java/com/ibm/fhir/audit/logging/api/configuration/ConfigurationTranslator.java:183]: make a list if you want formatting preserved

File: fhir-audit/src/main/java/com/ibm/fhir/audit/logging/mapper/impl/AuditEventMapper.java [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+/**
+ * This class adds support for AuditEvent
+ * https://www.hl7.org/fhir/r4/auditevent.html
+ */
+public class AuditEventMapper implements Mapper {
+
+    private static final Logger logger = java.util.logging.Logger.getLogger(AuditEventMapper.class.getName());
+    private static final String CLASSNAME = AuditEventMapper.class.getName();
+
+    //@formatter:off
```
Comment (lmsurpre) [inline fhir-audit/src/main/java/com/ibm/fhir/audit/logging/mapper/impl/AuditEventMapper.java:53]: personally not a huge fan of all the formatter comments, but i understand why you add them.  nothing to change here, but maybe eventually we can settle on a common style (include them or not) :-)

File: fhir-audit/src/main/java/com/ibm/fhir/audit/logging/mapper/impl/CADFMapper.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    private static final Logger logger = java.util.logging.Logger.getLogger(CLASSNAME);
+
+    // FHIR Operation To CADF
+    private static final Map<String, Action> FHIR_TO_CADF = new HashMap<String, Action>() {
+        private static final long serialVersionUID = 1L;
+        {
+            put("C", Action.create);
+            put("R", Action.read);
+            put("U", Action.update);
+            put("D", Action.delete);
```
Comment (lmsurpre) [inline fhir-audit/src/main/java/com/ibm/fhir/audit/logging/mapper/impl/CADFMapper.java:51]: do we ever get an "E" as input to this?

File: fhir-audit/src/main/java/com/ibm/fhir/audit/logging/mapper/impl/CADFMapper.java [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+                cadfEventOutCome = Outcome.success;
+            } else {
+                cadfEventOutCome = Outcome.failure;
+            }
+
+            event = new CadfEvent.Builder(
+                            logEntry.getContext().getRequestUniqueId() == null ? UUID.randomUUID().toString()
+                                    : logEntry.getContext().getRequestUniqueId(),
+                                    EventType.activity, logEntry.getTimestamp(),
+                                    FHIR_TO_CADF.getOrDefault(logEntry.getContext().getAction(), Action.unknown),
```
Comment (lmsurpre) [inline fhir-audit/src/main/java/com/ibm/fhir/audit/logging/mapper/impl/CADFMapper.java:142]: do we ever get an "E" as input to this?  if so, i guess we'll end up setting "Action.unknown" ?

File: fhir-audit/src/main/java/com/ibm/fhir/audit/logging/api/impl/NopService.java [hunk 9]
```diff
@@ -1,38 +1,42 @@
 /*
- * (C) Copyright IBM Corp. 2016,2019
+ * (C) Copyright IBM Corp. 2016,2020
  *
  * SPDX-License-Identifier: Apache-2.0
  */
 
-package com.ibm.fhir.audit.logging.impl;
+package com.ibm.fhir.audit.logging.api.impl;
```
Comment (lmsurpre) [inline fhir-audit/src/main/java/com/ibm/fhir/audit/logging/api/impl/NopService.java:7]: Why the `.api` in the package name?  Can we just use `com.ibm.fhir.audit.logging.impl` for the impl classes?

Comment (prb112) [reply fhir-audit/src/main/java/com/ibm/fhir/audit/logging/api/impl/NopService.java:7]: legacy -- I can change it... matched the existing package structure

Comment (lmsurpre) [reply fhir-audit/src/main/java/com/ibm/fhir/audit/logging/api/impl/NopService.java:7]: I pinged paul on the side and suggested that maybe we could even do away with the `logging` in the package name for all classes in this module.  My understanding is there is no other audit package other than logging and so this level seems totally pointless.

File: fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java [hunk 10]
```diff
@@ -498,6 +491,8 @@ private static AuditLogEntry populateAuditLogEntry(AuditLogEntry entry, HttpServ
                             .append("/")
                             .append(request.getRemoteHost()).toString());
         entry.setContext(new Context());
+
+        // TODO: Unify with the FHIRRequestContext
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java:495]: the one update I think you missed :-)

File: fhir-audit/src/main/java/com/ibm/fhir/audit/logging/api/impl/NopService.java (hunk 9 above)
Comment (prb112) [reply fhir-audit/src/main/java/com/ibm/fhir/audit/logging/api/impl/NopService.java:7]: Changed per discussion

File: fhir-audit/src/main/java/com/ibm/fhir/audit/logging/api/AuditLogServiceFactory.java (hunk 4 above)
Comment (prb112) [reply fhir-audit/src/main/java/com/ibm/fhir/audit/logging/api/AuditLogServiceFactory.java:2]: ✅

File: fhir-audit/src/main/java/com/ibm/fhir/audit/logging/api/configuration/ConfigurationTranslator.java (hunk 5 above)
Comment (prb112) [reply fhir-audit/src/main/java/com/ibm/fhir/audit/logging/api/configuration/ConfigurationTranslator.java:183]: fixed

File: fhir-audit/src/main/java/com/ibm/fhir/audit/logging/mapper/impl/AuditEventMapper.java (hunk 6 above)
Comment (prb112) [reply fhir-audit/src/main/java/com/ibm/fhir/audit/logging/mapper/impl/AuditEventMapper.java:53]: Sure - I hate hitting ctrl + f or having something accidentally formatted.

File: fhir-audit/src/main/java/com/ibm/fhir/audit/logging/mapper/impl/CADFMapper.java (hunk 8 above)
Comment (prb112) [reply fhir-audit/src/main/java/com/ibm/fhir/audit/logging/mapper/impl/CADFMapper.java:142]: `FHIR_TO_CADF.getOrDefault(logEntry.getContext().getAction(), Action.unknown),`
This ends up as an unknown. 

In the  output it looks like this...

Comment (prb112) [reply fhir-audit/src/main/java/com/ibm/fhir/audit/logging/mapper/impl/CADFMapper.java:142]:
```
CreateTime:1607706347894
{
    "action": "unknown",
    "eventTime": "2020-12-11 17:05:47.882",
    "eventType": "activity",
    "id": "336794e9-b5f9-4ac5-a00b-72aa7d9c5e0b",
    "outcome": "success",
    "typeURI": "http://schemas.dmtf.org/cloud/audit/1.0/event",
    "tags": [
    ],
    "attachments": [
        {
            "contentType": "application/json",
            "content": "rO0ABXQCFgp7CiAgICAicmVxdWVzdF91bmlxdWVfaWQiOiAiMzM2Nzk0ZTktYjVmOS00YWM1LWEwMGItNzJhYTdkOWM1ZTBiIiwKICAgICJhY3Rpb24iOiAiRSIsCiAgICAic3RhcnRfdGltZSI6ICIyMDIwLTEyLTExIDE3OjA1OjQ3LjUyNCIsCiAgICAiZW5kX3RpbWUiOiAiMjAyMC0xMi0xMSAxNzowNTo0Ny44ODIiLAogICAgImFwaV9wYXJhbWV0ZXJzIjogewogICAgICAgICJyZXF1ZXN0IjogImh0dHBzOi8vbG9jYWxob3N0Ojk0NDMvZmhpci1zZXJ2ZXIvYXBpL3Y0IiwKICAgICAgICAicmVxdWVzdF9zdGF0dXMiOiAyMDAKICAgIH0sCiAgICAiYmF0Y2giOiB7CiAgICAgICAgInJlc291cmNlc19yZWFkIjogMCwKICAgICAgICAicmVzb3VyY2VzX2NyZWF0ZWQiOiAwLAogICAgICAgICJyZXNvdXJjZXNfdXBkYXRlZCI6IDIKICAgIH0sCiAgICAiZXZlbnRfdHlwZSI6ICJmaGlyLWJ1bmRsZSIsCiAgICAiZGVzY3JpcHRpb24iOiAiRkhJUiBCdW5kbGUgcmVxdWVzdCIsCiAgICAibG9jYXRpb24iOiAiMDowOjA6MDowOjA6MDoxL2xvY2FsaG9zdCIKfQ=="
        }
    ],
```

File: fhir-audit/src/main/java/com/ibm/fhir/audit/logging/api/impl/NopService.java (hunk 9 above)
Comment (prb112) [reply fhir-audit/src/main/java/com/ibm/fhir/audit/logging/api/impl/NopService.java:7]: Discussed in chat

File: docs/src/pages/guides/FHIRServerUsersGuide.md (hunk 2 above)
Comment (prb112) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:1916]: Yeah - that's why I mentioned that it needs more work once we have a customer asking for a use-case

File: fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java (hunk 10 above)
Comment (prb112) [reply fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java:495]: Thanks, it's a online change, so easy to get in.

File: docs/src/pages/guides/FHIRServerUsersGuide.md [hunk 11]
```diff
@@ -27,7 +27,7 @@ permalink: /FHIRServerUsersGuide/
   * [4.8 Using local references within request bundles](#48-using-local-references-within-request-bundles)
   * [4.9 Multi-tenancy](#49-multi-tenancy)
```
Comment (tbieste) [inline docs/src/pages/guides/FHIRServerUsersGuide.md:28]: Section 4.9.2.1 (line 1481)
"serviceClassName" : "com.ibm.fhir.audit.logging.impl.DisabledAuditLogService",
Replace with No-op.

Comment (prb112) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:28]: hum I don't see it ...

Comment (prb112) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:28]: Weird my local copy doesn't show this

Comment (prb112) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:28]: OH this is an old commit

File: docs/src/pages/guides/FHIRServerUsersGuide.md [hunk 12]
```diff
@@ -1478,7 +1478,7 @@ The global configuration contains non-tenant specific configuration parameters (
             }
         },
         "audit": {
-            "serviceClassName" : "com.ibm.fhir.audit.logging.impl.DisabledAuditLogService",
```
Comment (prb112) [inline docs/src/pages/guides/FHIRServerUsersGuide.md:1481]: Changed here...

Comment (tbieste) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:1481]: Cool.  Thanks!

File: docs/src/pages/guides/FHIRServerUsersGuide.md (hunk 11 above)
Comment (tbieste) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:28]: ok, thanks!  Not sure why I was seeing that.  Looks good now.

**Review by tbieste — APPROVED** [review]

**Review by lmsurpre — APPROVED** [review]
LGTM.  We should add the old classnames used in the config under the Deprecated section of our 4.5.3 release notes.

**Comment by prb112** [comment]
> LGTM. We should add the old classnames used in the config under the Deprecated section of our 4.5.3 release notes.

I'll start working on that

File: fhir-audit/src/main/java/com/ibm/fhir/audit/configuration/ConfigurationTranslator.java [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+        }
+        if(!props.containsKey("security.protocol")) {
+            props.put("security.protocol", "SASL_SSL");
+        }
+        if(!props.containsKey("ssl.protocol")) {
+            props.put("ssl.protocol", "TLSv1.2");
+        }
+        if(!props.containsKey("ssl.enabled.protocols")) {
+            props.put("ssl.enabled.protocols", "TLSv1.2");
+        }
```
Comment (punktilious) [inline fhir-audit/src/main/java/com/ibm/fhir/audit/configuration/ConfigurationTranslator.java:175]: Perhaps constants would be better for these string values.

File: fhir-audit/src/main/java/com/ibm/fhir/audit/configuration/type/IBMEventStreamsType.java [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+                try (JsonReader jsonReader =
+                        JSON_READER_FACTORY.createReader(stringReader)) {
+                    JsonObject jsonObject = jsonReader.readObject();
+                    EventStreamsCredentials.Builder builder =
+                            EventStreamsCredentials.builder();
+
+                    if (jsonObject.get("api_key") != null) {
+                        String apiKey = jsonObject.getString("api_key");
+                        builder.apiKey(apiKey);
+                    }
```
Comment (punktilious) [inline fhir-audit/src/main/java/com/ibm/fhir/audit/configuration/type/IBMEventStreamsType.java:205]: String constants

**Pull request merged**